### PR TITLE
Fix/mob 1252 2 activity created in 1post

### PR DESCRIPTION
**Pull request by diego10bk**
- Use eCMS Rest Service to upload the photos in 2 steps:
+ Upload to tmp in server
+ Move the photo to the right place. 
- Use NSURLSession to send the requests. 
- The Request are all asynchronous

### CODE REVIEW

File: Sources/Shared/UI/Documents/DocumentsViewController.m [hunk 1]
```diff
@@ -227,6 +229,7 @@ - (void)dealloc
     _dicContentOfFolder = nil;
```
Comment (paristote) [inline Sources/Shared/UI/Documents/DocumentsViewController.m]: Some variables need to be released or set to `nil`
- `fileToApplyAction`
- `popoverPhotoLibraryController`
- `_stringForUploadPhoto`
- `popoverProperties`

File: Sources/Shared/UI/Documents/DocumentsViewController.m [hunk 2]
```diff
@@ -995,11 +998,17 @@ -(void)askToMakeFolderActions:(BOOL)createNewFolder{
 
 - (void)sendImageInBackgroundForDirectory:(NSString *)directory data:(NSData *)imageData
 {
-    [_filesProxy fileAction:kFileProtocolForUpload source:directory destination:nil data:imageData];
+    _filesProxy.delegate = self;
+    [_filesProxy uploadFile:imageData asFileName:[directory lastPathComponent] inFolder:fileToApplyAction.currentFolder ofDrive:fileToApplyAction.driveName];
     //Need to reload the content of the folder
+
     [self startRetrieveDirectoryContent];
```
Comment (paristote) [inline Sources/Shared/UI/Documents/DocumentsViewController.m]: This method is called after a successful upload (`didUploadImage`), so is it still necessary here?

File: Sources/Shared/UI/Documents/DocumentsViewController.m [hunk 3]
```diff
@@ -1068,15 +1077,16 @@ - (void)imagePickerController:(UIImagePickerController *)picker didFinishPicking
                 
                 //release the date formatter because, not needed after that piece of code
                 [dateFormatter release];
-                imageName = [NSString stringWithFormat:@"MobileImage_%@.png", tmp];
+                imageName = [NSString stringWithFormat:@"mobile_image_%@.png", tmp];
```
Comment (paristote) [inline Sources/Shared/UI/Documents/DocumentsViewController.m]: Just a question: why did you change the filename prefix?
Also, this prefix should be defined as a constant in defines.h and everywhere necessary.

File: Sources/Shared/UI/Social/MessageComposerViewController.m [hunk 4]
```diff
@@ -349,8 +350,14 @@ - (IBAction)onBtnSend:(id)sender
...
             FilesProxy *fileProxy = [FilesProxy sharedInstance];
+            NSString * photosFolderPath;
+            if (selectedSpace && selectedSpace.spaceId) {
+               photosFolderPath = [NSString stringWithFormat:@"%@/rest/private/jcr/%@/%@/Groups%@/Documents",[ApplicationPreferencesManager sharedInstance].selectedAccount.serverUrl,[ApplicationPreferencesManager sharedInstance].currentRepository, [ApplicationPreferencesManager sharedInstance].defaultWorkspace, selectedSpace.groupId];
+            } else {
+                photosFolderPath =  [NSString stringWithFormat:@"%@/Public", fileProxy._strUserRepository];
+            }
             
-            BOOL storageFolder = [fileProxy createNewFolderWithURL:[NSString stringWithFormat:@"%@/Public", fileProxy._strUserRepository] folderName:@"Mobile"];
+            BOOL storageFolder = [fileProxy createNewFolderWithURL:photosFolderPath folderName:@"Mobile"];
```
Comment (paristote) [inline Sources/Shared/UI/Social/MessageComposerViewController.m]: The folder name should be defined as a constant in defines.h, e.g.
`#define MOBILE_UPLOAD_DEST_FOLDER @"Mobile"`

File: Sources/Shared/UI/Social/MessageComposerViewController.m [hunk 5]
```diff
@@ -361,18 +368,18 @@ - (IBAction)onBtnSend:(id)sender
                 //release the date formatter because, not needed after that piece of code
                 [dateFormatter release];
 
-                fileAttachName = [NSString stringWithFormat:@"MobileImage_%@.png", fileAttachName];
+                fileAttachName = [NSString stringWithFormat:@"mobile_image_%@.png", fileAttachName];
                 fileAttachName = [fileAttachName stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
                 NSLog(@"uploading file: %@",fileAttachName);
                 
-                fileAttachURL = [NSString stringWithFormat:@"%@/Public/Mobile/%@", fileProxy._strUserRepository, fileAttachName];
+                fileAttachURL = [NSString stringWithFormat:@"%@/Mobile/%@",photosFolderPath,fileAttachName];
```
Comment (paristote) [inline Sources/Shared/UI/Social/MessageComposerViewController.m]: Following the comment above, it would become:
`[NSString stringWithFormat:@"%@/%@/%@", photosFolderPath, MOBILE_UPLOAD_DEST_FOLDER, fileAttachName];`

File: Sources/Shared/UI/Social/MessageComposerViewController.m [hunk 6]
```diff
@@ -416,6 +423,18 @@ - (IBAction)onBtnSend:(id)sender
...
 }
 
+- (void)sendImageInBackgroundForDirectory:(NSString *)directory data:(NSData *)imageData
+{
+    if (selectedSpace){
+        NSString * driverName = [selectedSpace.groupId stringByReplacingOccurrencesOfString:@"/" withString:@"."];
+        [[FilesProxy sharedInstance] uploadFile:imageData asFileName:[directory lastPathComponent] inFolder:@"Mobile" ofDrive:driverName];
+        
+    } else {
+        [[FilesProxy sharedInstance] uploadFile:imageData asFileName:[directory lastPathComponent] inFolder:@"Public/Mobile" ofDrive:@"Personal Documents"];
```
Comment (paristote) [inline Sources/Shared/UI/Social/MessageComposerViewController.m]: `Personal Documents` should be defined as 
`#define MOBILE_UPLOAD_PERSONAL_DRIVE @"Personal Documents"`

File: Sources/iPhone/Data Manager/Proxies/FilesProxy.m [hunk 7]
```diff
@@ -368,6 +372,87 @@ -(NSString *)fileAction:(NSString *)protocol source:(NSString *)source destinati
...
+     currentFolder= : the folder, within the drive, in which to move the file
+     fileName= : the name of the file (can be different than the original)
+     */
+    NSString * serverURL = [ApplicationPreferencesManager sharedInstance].selectedAccount.serverUrl;
+    
+    NSString * uploadId = [NSUUID UUID].UUIDString;
+    uploadId = [uploadId stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
+    NSString * boundary = [NSString stringWithFormat:@"-----%@",uploadId];
+    
+    NSString * postRESTURL = [NSString stringWithFormat:@"%@/portal/rest/managedocument/uploadFile/upload?uploadId=%@&action=upload", serverURL, uploadId];
```
Comment (paristote) [inline Sources/iPhone/Data Manager/Proxies/FilesProxy.m]: URL endpoint should be defined in defines.h with other `DOCUMENT_...` constants, e.g.
`DOCUMENT_UPLOAD_SERVICE_PATH @"/portal/rest/managedocument/uploadFile/upload"`
and
`DOCUMENT_SAVE_SERVICE_PATH @"/portal/rest/managedocument/uploadFile/control"`

File: Sources/iPhone/Data Manager/Proxies/FilesProxy.m [hunk 8]
```diff
@@ -368,6 +372,87 @@ -(NSString *)fileAction:(NSString *)protocol source:(NSString *)source destinati
...
+            NSMutableURLRequest* request = [[NSMutableURLRequest alloc] init];
+            [request setURL:[NSURL URLWithString:saveRESTURL]];
+            [request setHTTPMethod:@"GET"];
+            NSURLSessionDataTask *dataTask = [aSession dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+                NSUInteger statusCode = [((NSHTTPURLResponse*) response) statusCode];
+                if(statusCode >= 200 && statusCode < 300) {
+                    if (delegate && [delegate respondsToSelector:@selector(fileProxy:didUploadImage:)]) {
+                        [delegate fileProxy:self didUploadImage:YES];
+                    }
+                }
```
Comment (paristote) [inline Sources/iPhone/Data Manager/Proxies/FilesProxy.m]: if the operation is not successful, call the delegate method with param `NO`.

File: Sources/Shared/UI/Documents/DocumentsViewController.m [hunk 9]
```diff
@@ -995,11 +998,17 @@ -(void)askToMakeFolderActions:(BOOL)createNewFolder{
...
+    [_filesProxy uploadFile:imageData asFileName:[directory lastPathComponent] inFolder:fileToApplyAction.currentFolder ofDrive:fileToApplyAction.driveName];
     //Need to reload the content of the folder
+
     [self startRetrieveDirectoryContent];
 }
-
+-(void) fileProxy:(FilesProxy *)fileProxy didUploadImage:(BOOL)finished {
+    if (finished) {
+        [self startRetrieveDirectoryContent];
+    }
```
Comment (paristote) [inline Sources/Shared/UI/Documents/DocumentsViewController.m]: `else {` and show an error message to the user `Upload failed. Please try again later.`

File: Sources/Shared/UI/Documents/DocumentsViewController.m (hunk 3 above)
Comment (diego10bk) [inline Sources/Shared/UI/Documents/DocumentsViewController.m]: Like in extension, we have the same regles of file name to use eCMS to upload, hasn't it?

File: Sources/Shared/UI/Documents/DocumentsViewController.m [hunk 10]
```diff
@@ -995,11 +1001,17 @@ -(void)askToMakeFolderActions:(BOOL)createNewFolder{
...
-    //Need to reload the content of the folder
-    [self startRetrieveDirectoryContent];
+    _filesProxy.delegate = self;
+    [_filesProxy uploadFile:imageData asFileName:[directory lastPathComponent] inFolder:fileToApplyAction.currentFolder ofDrive:fileToApplyAction.driveName];
+}
+-(void) fileProxy:(FilesProxy *)fileProxy didUploadImage:(BOOL)finished {
+    if (finished) {
+        [self startRetrieveDirectoryContent];
+    } else {
+        UIAlertView * alert = [[UIAlertView alloc] initWithTitle:Localize(@"Upload failed") message:Localize(@"Please try again later") delegate:nil cancelButtonTitle:Localize(@"Close") otherButtonTitles:nil];
```
Comment (paristote) [inline Sources/Shared/UI/Documents/DocumentsViewController.m]: Add the `autorelease` statement for the alert.

File: Sources/Shared/Data Manager/Proxies/Social/SocialPostActivity.m [hunk 11]
```diff
@@ -98,11 +98,16 @@ -(void)postActivity:(NSString *)message fileURL:(NSString*)fileURL fileName:(NSS
         ApplicationPreferencesManager *serverPM = [ApplicationPreferencesManager sharedInstance];
         activity.type = @"DOC_ACTIVITY"; 
         
+        NSString* docPath;
+        if (space){
+            docPath = [NSString stringWithFormat:@"/Groups%@/Documents/Mobile/%@",space.groupId,fileName];
+        } else {
+            docPath = [NSString stringWithFormat:@"%@/Public/Mobile/%@",serverPM.userHomeJcrPath,fileName];
```
Comment (paristote) [inline Sources/Shared/Data Manager/Proxies/Social/SocialPostActivity.m]: You can use `MOBILE_UPLOAD_DEST_FOLDER` instead of `Mobile` in the 2 paths above.

File: Sources/Shared/UI/Social/MessageComposerViewController.m (hunk 5 above)
Comment (paristote) [inline Sources/Shared/UI/Social/MessageComposerViewController.m]: Please use `MOBILE_UPLOAD_DEST_FOLDER`

File: Sources/Shared/UI/Documents/DocumentsViewController.m [hunk 12]
```diff
@@ -573,8 +579,8 @@ - (void)showImagePickerForAddPhotoAction:(UIImagePickerController *)picker {
 
 - (void)showErrorForFileAction:(NSString *)errorMessage {
     [self hideLoader:NO];
-    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:Localize(@"FileError") message:Localize(errorMessage) delegate:self 
-                                          cancelButtonTitle:Localize(@"OK") otherButtonTitles:nil, nil];
+    UIAlertView *alert = [[[UIAlertView alloc] initWithTitle:Localize(@"FileError") message:Localize(errorMessage) delegate:self
+                                          cancelButtonTitle:Localize(@"OK") otherButtonTitles:nil, nil] autorelease];
     [alert show];
     [alert release];
```
Comment (paristote) [inline Sources/Shared/UI/Documents/DocumentsViewController.m]: Pay attention! The object will be released twice!

File: Sources/Shared/UI/Documents/DocumentsViewController.m (hunk 10 above)
Comment (paristote) [inline Sources/Shared/UI/Documents/DocumentsViewController.m]: Still need to release this object.

**Comment by paristote** [comment]
Fixed by b32459c.
